src/Execute/DeviceViewProvider.ts

```diff
@@ -0,0 +1,208 @@
+
+enum NodeType {
+  /**
+   * A Execution system that contains multiple Devices.
+   */
+  executionManager,
+  /**
+   * A single device dedicated to certain device.
+   */
+  device,
+  /**
+   * A Executor that run on certain device.
+   */
+  executor,
+  /**
+   * Node for empty item.
+   * No icon for this.
+   */
+  none,
+}
+
+export class DeviceViewProvider implements vscode.TreeDataProvider<DeviceViewNode> {
+  executionManagerMap: ExecutionManagerMap = {};
+  private _onDidChangeTreeData: vscode.EventEmitter<DeviceTreeView> =
+      new vscode.EventEmitter<DeviceViewNode>();
+  readonly onDidChangeTreeData: vscode.Event<DeviceTreeView> = this._onDidChangeTreeData.event;
+  getTreeItem(_element: DeviceViewNode): vscode.TreeItem|Thenable<vscode.TreeItem> {
+    return _element;
+  }
+  getChildren(_element?: DeviceViewNode|undefined): vscode.ProviderResult<DeviceViewNode[]> {
+    return this.getNodes(_element);
+  }
+
+  getNodes(_element: DeviceViewNode|undefined): DeviceViewNode[] {
+    const rtnList: DeviceViewNode[] = [];
+    if (!_element) {
+      for (const managerName of executionManagerList) {
+        rtnList.push(new DeviceViewNode(
+            managerName, vscode.TreeItemCollapsibleState.Expanded, NodeType.executionManager,
+            managerName));
+      }
+    } else if (_element.type === NodeType.executionManager) {
+      const deviceList = this.executionManagerMap[_element.managerName].allDevices;
+      for (const device of deviceList) {
+        rtnList.push(new DeviceViewNode(
+            device.name, vscode.TreeItemCollapsibleState.Collapsed, NodeType.device,
+            _element.managerName));
+      }
+    } else if (_element.type === NodeType.device) {
+      const device = this.executionManagerMap[_element.managerName].findDevice(_element.labal);
+      const executorList = device ?.availableExecutors;
+      if (executorList) {
+        for (const executor of executorList) {
+          // Currently, No Executor name for this, so need to update Executor name first.
+          rtnList.push(new DeviceViewNode(
+              executor.name(), vscode.TreeItemCollapsibleState.None, NodeType.executor,
+              _element.managerName));
+        }
+      }
+    }
+    if (rtnList.length === 0) {
+      rtnList.push(new DeviceViewNode(
+          'No subtree available', vscode.TreeItemCollapsibleState.None, NodeType.none, ''));
+    }
+    return rtnList;
+  }
+
+  getDevicesPromise(cmd: Command, deviceSpec: DeviceSpec): Promise<Device[]> {
+    return new Promise<Device[]>((resolve, _reject) => {
+      cp.exec(cmd.str(), (err, stdout, _stderr) => {
+        if (err) {
+          return;
+        }
+        const devices: Device[] = [];
+        const deviceNames = String(stdout).split('\n').filter(element => element);
+        for (const deviceName of deviceNames) {
+          devices.push(new Device(deviceName, deviceSpec));
+        }
+        resolve(devices);
+      });
```

Comment:
this should be changed to use 'spawn' with {shell: false}. this change will be updated after #1037 1037